### PR TITLE
Add Task for publishing tekton pipeline images + yaml

### LINKED ISSUE #528: Create `Task` for building and tagging Tekton Pipelines release images
## Expected Behavior
We should be testing and releasing Pipelines with Pipelines.

We should have a `Task` (and related Resources) defined that we can use to:
- Build all images 
- Tag the images appropriately

It should be possible to use this for both:
- Manually creating a versioned release (including updating "latest" tag)
- Creating nightly releases

### The Task

* Inputs:
  - git repo
* Outputs:
  - gcr images (for all Tekton images)
* Parameters:
- Version to use (maybe absence indicates nightly release?)

## Actual Behavior

We currently use [prow](https://github.com/knative/test-infra/tree/master/ci/prow) and [knative release scripts](https://github.com/knative/build-pipeline/blob/master/hack/release.md) for releasing.

## Additional Info

* See [0.1 release.yaml](https://github.com/knative/build-pipeline/releases/download/v0.1.0/release.yaml) to see all images required
* Actually triggering this is not in scope for this issue, it's fine to test manually (we'll be linking these together in a later issue)
* PR should include docs that explain how the Task works

### LINKED ISSUE #529: Create `Task` for publishing Tekton Pipeline release yaml
## Expected behaviour

We should be testing and releasing Pipelines with Pipelines.

We should have a `Task` (and related Resources) defined that we can use to:
- Generate the `release.yaml` (including the correct image names)
- Upload the `release.yaml` to a gcs bucket

### The Task

* Inputs:
  - git repo
  - gcr images (for all Tekton images - we can assume these have been built by something upstream, specifically this is going to be #528)
* Outputs:
  - gcs bucket (where the release.yaml will be written)
* Parameters:
- ?

## Actual Behavior

We currently use [prow](https://github.com/knative/test-infra/tree/master/ci/prow) and [knative release scripts](https://github.com/knative/build-pipeline/blob/master/hack/release.md) for releasing.

## Additional Info

* Blocked by https://github.com/knative/build-pipeline/issues/216 (we have no way to get the digests of previously built images) - we could implement this assuming we have support for digests if we want (the field does exist in the PipelineResource)
* See [0.1 release.yaml](https://github.com/knative/build-pipeline/releases/download/v0.1.0/release.yaml) to see all images required
* Actually triggering this is not in scope for this issue, it's fine to test manually (we'll be linking these together in a later issue)
* PR should include docs that explain how the Task works

### PR DESCRIPTION
**Pull request by bobcatfish**
# Changes

Add a `Task` which invokes `ko` to build and publish all images and yaml
config required for installing Tekton Pipelines.

This Task will:
* Build and publish the "base image" using Kaniko
* Generate a .ko.yaml
* Invoke ko to build/publish images and generate a release.yaml
* Parse the release.yaml for built images; ensuring that the expected
  images were built (and no more)
* Tag the built images with the correct version and also tag in all
  regions (us, asia, eu)

This should be the same functionality that could previously be seen in
https://github.com/tektoncd/pipeline/blob/master/hack/release.sh
(which used
https://github.com/knative/test-infra/blob/master/scripts/release.sh).
We can remove release.sh once we have completed #530 as well.

Some functionality has been implemented in a python script, which has
its own tests. Since it is currently difficult to update the pull
request test logic to do additional things (such as run python unit
tests), I'm hoping we are okay with waiting until #532 to add
automatic running of these tests).

Fixes #528
Fixes #529

# Submitter Checklist

These are the criteria that every PR should meet, please check them off as you
review them:

- [x] Includes [tests](https://github.com/tektoncd/pipeline/blob/master/CONTRIBUTING.md#principles) (if functionality changed/added)
- [x] Includes [docs](https://github.com/tektoncd/pipeline/blob/master/CONTRIBUTING.md#principles) (if user facing)
- [x] Commit messages follow [commit message best practices](https://github.com/tektoncd/pipeline/blob/master/CONTRIBUTING.md#commit-messages)

_See [the contribution guide](https://github.com/tektoncd/pipeline/blob/master/CONTRIBUTING.md)
for more details._

# Release Notes
```
n/a
```

### CODE REVIEW

File: tekton/koparse/koparse.py [hunk 1]
```diff
@@ -0,0 +1,116 @@
...
+        base: The built images will be expected to start with this string,
+            other images will be ignored
+        path: The path to the file (release.yaml) that will contain the built images
+    Returns:
+        list of the images parsed from the file
+    """
+    images = []
+    with open(path) as f:
+        for line in f:
+            match = re.search(base + ".*@sha256:[0-9a-f]*", line)
```
Comment (tsandall) [inline tekton/koparse/koparse.py]: u shuld use DIGGEST_MARKER here ;)

File: tekton/publish.yaml [hunk 2]
```diff
@@ -0,0 +1,183 @@
...
+    - |
+      set -e
+      set -x
+      # TODO(#631): this is a hack to make the go binary available in a container that has gcloud
+      mkdir -p /workspace/golang
+      cp /usr/local/go/bin/go /workspace/golang/go
+      cp -R /usr/local/go /workspace/golang/localgo
+
+  - name: ensure-release-dirs-exist
+    image: busybox
```
Comment (pivotal-nader-ziada) [inline tekton/publish.yaml]: when the bucket was empty, I got the error
```
error exit status 1; cmd_output CommandException: No URLs matched: gs://pipeline-test-123456/*\n","stacktrace":"main.main\n\t/Users/pivotal/go/src/github.com/tektoncd/pipeline/cmd/gsutil/main.go:94\nruntime.main\n\t/usr/local/Cellar/go/1.12/libexec/src/runtime/proc.go:200"
```

but that might be a bug in the resource itself

File: hack/release.md [hunk 3]
```diff
@@ -1,5 +1,7 @@
 # Creating a new Tekton Pipeline release
 
+**Note: we are transitioning to a Pipelines based test and release process, see [tekton/README.md](tekton.README.md).**
```
Comment (pivotal-nader-ziada) [inline hack/release.md]: (tekton/README.md) ?

Comment (vdemeester) [reply hack/release.md]: `../README.md` maybe ?

File: tekton/README.md [hunk 4]
```diff
@@ -0,0 +1,71 @@
...
+you create the `TaskRun`. If that default service account is able to push to your image registry,
+you are good to go. Otherwise, you need to use [a secret annotated with your docker registry
+credentials](https://github.com/tektoncd/pipeline/blob/master/docs/auth.md#basic-authentication-docker).
+
+TODO(#631) Ensure that we are supporting folks using credentials other than the cluster defaults; not
+sure how this will play out with publishing to our prod registry!
+
+#### Production credentials
+
+TODO(dlorenc, bobcatfish): We need to setup a group which users can be added to, as well as guidelines
```
Comment (vdemeester) [inline tekton/README.md]: :100: :heart_eyes_cat:

File: tekton/koparse/koparse.py [hunk 5]
```diff
@@ -0,0 +1,116 @@
+#!/usr/bin/env python2.7
```
Comment (vdemeester) [inline tekton/koparse/koparse.py]: ohh fancy

Comment (vdemeester) [reply tekton/koparse/koparse.py]: Why not `python3` though :trollface:

Comment (bobcatfish) [reply tekton/koparse/koparse.py]: lol you raise a good point @vdemeester XD

**Comment by bobcatfish** [comment]
```
I0319 00:22:06.449] =============================
I0319 00:22:06.461] ----------------------------------------------
I0319 00:22:06.461] ---- Checking links in the markdown files ----
I0319 00:22:06.462] ----------------------------------------------
I0319 00:22:14.228] hack/release.md
I0319 00:22:14.228] 	ERROR	tekton.README.md
I0319 00:22:14.228] 		Stat hack/tekton.README.md: no such file or directory
```

lol whaaaaat

File: tekton/publish.yaml [hunk 6]
```diff
@@ -0,0 +1,183 @@
...
+    image: golang
+    env:
+    - name: GOBIN
+      value: /workspace
+    command: ["go"]
+    args:
+    - "get"
+    - "github.com/google/go-containerregistry/cmd/ko"
+
+  # TODO(#631): publish a `ko` image (which has golang)
```
[the remark this answers is not in the document]
Comment (bobcatfish) [reply tekton/publish.yaml]: Yeah definitely!! I was thinking of leaving that for #631 but you're probably right, it's worth doing something better than this in the meantime XD

File: tekton/publish.yaml (hunk 2 above)
Comment (bobcatfish) [reply tekton/publish.yaml]: oh interesting - so you'd created the bucket but nothing was in it? ill try that out too

File: hack/release.md (hunk 3 above)
Comment (bobcatfish) [reply hack/release.md]: :sweat_smile:

**Comment by bobcatfish** [comment]
I haven't finished responding to all the feedback yet but I did:
- Switch to Python 3
- Add a Dockerfile for the ko image (and a Task to build it!)

So those bits are ready for another look :D

File: tekton/koparse/test_koparse.py [hunk 7]
```diff
@@ -0,0 +1,71 @@
...
+
+import koparse
+
+
+IMAGE_BASE = "gcr.io/knative-releases/github.com/knative/build-pipeline/cmd/"
+PATH_TO_TEST_RELEASE_YAML = os.path.join(os.path.dirname(
+    os.path.abspath(__file__)), "test_release.yaml")
+PATH_TO_WRONG_FILE = os.path.join(os.path.dirname(
+    os.path.abspath(__file__)), "koparse.py")
+BUILT_IMAGES = [
```
Comment (vdemeester) [inline tekton/koparse/test_koparse.py]: We need `cmd/entrypoint` too :angel:

File: tekton/koparse/test_koparse.py [hunk 8]
```diff
@@ -0,0 +1,71 @@
...
+    "gcr.io/knative-releases/github.com/knative/build-pipeline/cmd/creds-init@sha256:67448da79e4731ab534b91df08da547bc434ab08e41d905858f2244e70290f48",
+    "gcr.io/knative-releases/github.com/knative/build-pipeline/cmd/git-init@sha256:7d5520efa2d55e1346c424797988c541327ee52ef810a840b5c6f278a9de934a",
+    "gcr.io/knative-releases/github.com/knative/build-pipeline/cmd/nop@sha256:3784d6b8f73043a29d2c1d6196801bee46fe808fbb94ba4fd21ca52dce503183",
+    "gcr.io/knative-releases/github.com/knative/build-pipeline/cmd/bash@sha256:d55917ef5c92627027e3755bfc577fbfa2fb783cccfb13a98632cb6ba6088cd6",
+    "gcr.io/knative-releases/github.com/knative/build-pipeline/cmd/gsutil@sha256:421a261436e16af4057b4a069fdae8a5aca6e37269952209ad9932a774aa0003",
+    "gcr.io/knative-releases/github.com/knative/build-pipeline/cmd/controller@sha256:bdc6f22a44944c829983c30213091b60f490b41f89577e8492f6a2936be0df41",
+    "gcr.io/knative-releases/github.com/knative/build-pipeline/cmd/webhook@sha256:cca7069a11aaf0d9d214306d456bc40b2e33e5839429bf07c123ad964d495d8a",
+]
+EXPECTED_IMAGES = [
+    "gcr.io/knative-releases/github.com/knative/build-pipeline/cmd/kubeconfigwriter",
```
Comment (vdemeester) [inline tekton/koparse/test_koparse.py]: here too :sweat_smile:

File: tekton/publish.yaml [hunk 9]
```diff
@@ -0,0 +1,183 @@
...
+    image: busybox
+    command:
+    - /bin/sh
+    args:
+    - -ce
+    - |
+      set -e
+      set -x
+
+      cat <<EOF > /workspace/go/src/github.com/tektoncd/pipeline/.ko.yaml
```
[the remark this answers is not in the document]
Comment (vdemeester) [reply tekton/publish.yaml]: @bobcatfish what is the default base image if we don't override ?

[the remark this answers is not in the document]
Comment (pivotal-nader-ziada) [reply tekton/publish.yaml]: if you don't specify any base image, `ko` will use [distroless](https://github.com/GoogleContainerTools/distroless)

**Comment by pivotal-nader-ziada** [comment]
@bobcatfish PR #643 update the release script, you may want to include that change in your task

File: tekton/publish.yaml (hunk 9 above)
[the remark this answers is not in the document]
Comment (bobcatfish) [reply tekton/publish.yaml]: adding a comment about the default!

File: tekton/publish.yaml (hunk 2 above)
Comment (bobcatfish) [reply tekton/publish.yaml]: hm okay so it seems like if you use the `dir` flag and the bucket is empty, you get this error:

```
Error executing command "gsutil cp -r gs://christies-empty-bucket/* /workspace/bucket" ; error exit status 1; cmd_output CommandException: No URLs matched: gs://christies-empty-bucket/*
"   
  stacktrace:  "main.main
	/Users/christiewilson/Code/go/src/github.com/tektoncd/pipeline/cmd/gsutil/main.go:94
runtime.main
	/usr/local/Cellar/go/1.10.3/libexec/src/runtime/proc.go:198"   
  ts:  1553120763.622061   
```

My feeling is that this is a bug in the resource; I think we should be able to handle checking out an empty bucket that we intend to put content into.

Comment (bobcatfish) [reply tekton/publish.yaml]: Created https://github.com/tektoncd/pipeline/issues/646

File: tekton/publish.yaml (hunk 9 above)
[the remark this answers is not in the document]
Comment (bobcatfish) [reply tekton/publish.yaml]: added entrypoint image :D

File: tekton/koparse/test_koparse.py (hunk 7 above)
Comment (bobcatfish) [reply tekton/koparse/test_koparse.py]: actually the test data (which is the 0.1 release) doesnt have the entrypoint image! 😇 

but this reminds me, i forgot to address @pivotal-nader-ziada 's comment and add the entrypoint image to the ko.yaml so I'll do that now 🤦‍♀️ 😅

**Comment by bobcatfish** [comment]
Okay I think I've addressed all comments @pivotal-nader-ziada @vdemeester , ready for another look :D

**Comment by bobcatfish** [comment]
```
I0321 17:57:39.418]  detail: u"Deploy error: Not all instances running in IGM after 25.116998368s. Expect 1. Current errors: [ZONE_RESOURCE_POOL_EXHAUSTED_WITH_DETAILS]: Instance 'gke-kpipeline-e2e-cls110-default-pool-7aa6058d-24gs' creation failed: The zone 'projects/tekton-prow-5/zones/us-west1-c' does not have enough resources available to fulfill the request.  '(resource type:compute)'. - ; ."
```

Well my current strategy of ignoring this error and hoping it goes away doesn't seem to be working XD

**Comment by bobcatfish** [comment]
I've manually applied the `us-central1` change from #651 so trying this again:

/test pull-tekton-pipeline-integration-tests

**Comment by bobcatfish** [comment]
```
timeout_test.go:222: Error waiting for TaskRun run-giraffe to finish: timed out waiting for the condition
```

o.o

/test pull-tekton-pipeline-integration-tests

**Comment by bobcatfish** [comment]
lol so many tests failing with `race detected during execution of test` 🏎 interesting...

**Pull request merged**